src/maskfile.c

```diff
@@ -0,0 +1,181 @@
+	// cp2 is used to convert \, back into plain , characters.
+	while (cp && i < MAX_NUM_CUST_PLHDR) {
+		char mask_param[512];
+		int len;
+		if (cp-cp1 > sizeof(mask_param))
+			len = sizeof(mask_param)-1;
+		else
+			len = cp-cp1;
+		strnzcpy(mask_param, cp1, len+1);
+		// we have to eat any escaped commas
+		cp2 = strstr(mask_param, "\\,");
+		while (cp2) {
+			memmove (cp2, cp2+1, strlen(cp2));
+			cp2 = strstr(cp2, "\\,");
+		}
+		// ok, now set this param up into the proper custom_mask value.
+		options.custom_mask[i++] = str_alloc_copy(mask_param);
+		// move cp1 to the next mask_parm, OR the real mask.
+		cp1 = cp+1;
+		// find the next param if there is one.
+		cp = strchr(cp1, ',');
+		while (cp && cp[-1] == '\\')
+			cp = strchr(&cp[1], ',');
+	}
+	// Ok, now cp1 should point to the real mask.
+	// simply eat any escaped commas
+	strnzcpy(out, cp1, outlen);
+	cp2 = strstr(out, "\\,");
+	while (cp2) {
+		memmove (cp2, cp2+1, strlen(cp2));
+		cp2 = strstr(cp2, "\\,");
+	}
+	return out;
+}
+
+void maskfile_hybrid_fix_state()
+{
+//	fprintf(stderr, "maskfile_hybrid_fix_state() linenum=%d local_linenum=%d\n", linenum, local_linenum);
+//	linenum = local_linenum;
+
+//	fprintf(stderr, "maskfile_hybrid_fix_state() linenum=%d\n", linenum);
+}
+
+int maskfile_restore_state_hybrid(const char *sig, FILE *fp) {
+	if (!strncmp(sig, "MSKF-v1", 7)) {
+		fscanf(fp, "%d\n", &linenum);
+//		fprintf(stderr, "maskfile_restore_state_hybrid() linenum=%d\n", linenum);
+	}
+	return 0;
+}
+
+static void maskfile_save_mode(FILE *fp) {
+//	fprintf(stderr, "maskfile_save_mode() linenum=%d\n", linenum);
+	if (linenum)
+		fprintf(fp, "MSKF-v1\n%d\n", linenum-1);
+}
+
+// this is like a do_crack.  yes, it needs a lot of work, but this is
+// a PoC that gets the water warmed up.
+void do_maskfile_crack(struct db_main *database, const char *fname) {
+	FILE *in = fopen(fname, "r");
+	char hBuf[512], linebuf[512];
+	int i;
+	if (!in) {
+		fprintf (stderr, "Error opening hc-mask file %s\n", fname);
+		exit(0);
+	}
+	rec_init_hybrid(maskfile_save_mode);
+	mask_crk_init(database);
+	for (i = 1; i < linenum; ++i) {
+		fgetl(linebuf, sizeof(linebuf)-1, in);
+		log_event("Resuming, skipping already done line %s", linebuf);
+	}
+	fgetl(linebuf, sizeof(linebuf)-1, in);
+	if (options.verbosity > VERB_LEGACY && i > 1)
+		fprintf(stderr, "Resuming maskfile: %s line: %d  Mask: %s\n",
+		                 fname, i, linebuf);
+	++linenum;
+	while (!feof(in)) {
+		maskfile_producemask(hBuf, sizeof(hBuf), linebuf);
+		if (*hBuf == 0) {
+			fgetl(linebuf, sizeof(linebuf)-1, in);
+			continue;
+		}
+		log_event("Processing maskfile line: %s", linebuf);
+		if (options.verbosity > VERB_LEGACY)
+		fprintf(stderr, "Processing maskfile: %s line: %d  Mask: %s\n",
+		                 fname, linenum, linebuf);
+		mask_init(database, hBuf);
+		mask_reset();
+		if (do_mask_crack(NULL))
+			break;
+		fgetl(linebuf, sizeof(linebuf)-1, in);
+		++linenum;
```

Comment:
I can imagine (but am not sure) we could feed each line to mask mode and see if it parses ok, then just reset without actually calling any do_crack